### DISCUSSION TITLE
H3S vc-relax weird behavior - Free Energy not decreasing?

### DISCUSSION THREAD

**Discussion by adenchfi** [SSCHAcode/python-sscha#40] [Q&A] [answered]
Hello, 
The SSCHA is an amazing idea! Thank you all for developing it. 

I tried running the H3S example on the Bebop cluster at ANL, where my main change was running the calculations at fixed pressure instead of fixed volume, and changing the # of kpoints. I seem to get strange results: 

![image](https://user-images.githubusercontent.com/10607496/111040457-a2b79380-83f8-11eb-86ea-22c93687c469.png)

The free energy has that strange jump, and then the code crashes with a segfault without any other information. I suspected it could be because of the Kong Liu effective sample size parameter getting smaller, so I tried to adapt the code to use the Relax module used for the LaH10 example. Here's the modified code: 
[H3S_initphon.txt](https://github.com/SSCHAcode/python-sscha/files/6135339/H3S_initphon.txt)
[SSCHArelax.txt](https://github.com/SSCHAcode/python-sscha/files/6135337/SSCHArelax.txt)

The H3S_initphon script seems to work perfectly fine. The SSCHA relax script has the following output behavior:
![all4output](https://user-images.githubusercontent.com/10607496/111040566-1659a080-83f9-11eb-8166-f00dcbc379c1.png)

The free energy has this strange behavior, and those large 'wide' regions are actually huge oscillations between steps. The code runs for about 6 iterations of the relax process, and then crashes. 
It segfaults with the message `/var/spool/slurmd/job1999230/slurm_script: line 23:  1000 Segmentation fault      (core dumped) python3 SSCHArelax.py > relaxoutput.out`. It doesn't appear to be a memory issue as it segfaults at the same spot whether or not I use more nodes. 

If I change the # of kpoints, it changes how many iterations the code goes for before segfaulting. For 2x2x2 kpoints, it ran for 4 iterations before segfaulting. 

The only thing I can think of is that the code isn't fully supporting of fixed pressure (as opposed to fixed-volume) calculations? But I'm not sure.

Looking forward to hearing back,
Adam

**Comment by mesonepigreco** [comment, marked as answer]
Hi, the first case of minimization is actually very good. The jump you see in the free energy is normal and simply a consequence of the fact you generated a new ensemble of ionic configurations. In this way, the free energy (which contains a term computed averaging the energy of the extracted configurations) will change in the new ensemble, but, as you see, remains compatible with the previous value within the stochastic error.

Actually, in the first case, I think the code converged (both gradients reached zero within their stochastic error). It is quite surprising it ended with a seg. fault. can you attach the full output so we can try to spot what went wrong? 
Did the code print the dynamical matrix of the second population (should be named dyn_pop2_X)?

In the second case, if you optimize at a fixed pressure, you are minimizing the Gibbs free energy. The code is printing the Helmoltz free energy, so you should probably add a PV term to the free energy to see that one. Since it is a variable cell relaxation, your volume changes abruptly after each ensemble to optimize the Gibbs free energy. This is probably giving the jumps in the free energy. In particular, it seems you are optimizing the cell at a bigger pressure than the starting one, this means your volume is getting smaller after each population (to optimize the target pressure). Indeed, the Helmoltz free energy increases a lot when the volume decreases.
However check the pressure printed in output by the code:

```
cat output | grep "STRESS" -A 5
```

It will print the stress tensor after each minimization. You can check if it is correctly converging to the pressure value you choose.

Also in the second case, if you look at both the gradient and the Kong-Liu (Neff / N0) ratio, the code seems to be converged after almost every iteration (apart from the first). You recognize this because the Neff value reaches 1/2 of the initial value (N0) only in the first population. In this case the code halts because the ensemble is no more good, and a new ensemble is generated. However, in the next populations, the code stops when Neff > N0 / 2, this means that it stops because the gradient converged to 0 within the stochastic error (the simulation converged). You can check this looking at the gradient plot. Therefore, also here the code probably stopped because it reached the correct value of the pressure. 
Again, it is a bit surprising this seg. fault error, If you can attach the output of relaxoutput.out we can check what are the last things the code does before crashing.
Bests,
Lorenzo

**Comment by adenchfi** [comment]
Hello,

Thank you for the thorough response! I learned a few things from it. Yes, I am getting "/var/spool/slurmd/job1999230/slurm_script: line 23:  1000 Segmentation fault      (core dumped) python3 SSCHArelax.py > relaxoutput.out" at the end, after AMASS: {}  {} is printed. 

Extracting the pressure via the cat STRESS command, it appears the pressure I have started at 219 GPa (I tried relaxing it to 210 GPa), but it increases during each SSCHA step, up to 420 GPa or so. Hm. That is strange. My input desired pressure is 210 GPa, and it stopped once the code hit ~ 420 GPa. Maybe there's something accidentally doubling the target pressure? 

Attached is the output text file for my 2nd case, along with input files for the phonon calc (H3S_initphon.py; this ran fine) ahd the SSCHA relaxation (SSCHArelax.py). 

[zipoutput.zip](https://github.com/SSCHAcode/python-sscha/files/6160749/zipoutput.zip)

Perhaps me specifying both 'press: 2100' in the SCF file input, and "target_press=2100" in the relax.vc_relax call resulted in a doubling to 420 GPa? While I wait for a reply, I'll try running things without the "press: 2100" in the SCF file input.

**Reply by mesonepigreco** [reply]
Just a quick reply, 
the pressure given to the vc_relax function of pyhton-sscha is in GPa already (not kbar as in quantum espresso). Thus you are trying to relax it to 2100 GPa! That is why the pressure is increasing.
The simulation stops at 420 GPa just because it reaches the maximum number of population you specified.
In fact it does 6 recalculations, as you told the code to stop after 6 populations:

```
relax = sscha.Relax.SSCHA(minimizer, ase_calculator = espresso_calc,
                         N_configs = 200,
                         max_pop = 6,  
                         save_ensemble = True,
                          cluster = None)
```

 Remember also that using the constant pressure relaxation could be dangerous if you use not well converged basis set of plane waves with ab initio code. In fact, the pressure given in output by quantum espresso pw.x has an offset if the basis set is not perfectly converged. 
This offset is automatically corrected if you optimize at fixed volume  (as you remove a uniform pressure to avoid a global volume rescale, which includes also this offset). The minimization at fixed volume does not mean fixed unit cell: all the lattice parameters are optimized, but the total volume is kept constant. 
This is particularly true if you want to optimize very small values of the pressures, like 1 GPa.  

Bests,
Lorenzo

**Reply by adenchfi** [reply]
Hello Lorenzo,

Thanks so much for the quick reply! That is most likely what's going on, or at least contributes to it. I'll fix the things you pointed out and try running things again. 

I see; so is there no practical difference between a fixed volume and fixed pressure calculation? Or am I misinterpreting your statement?

The planewave cutoffs are higher than recommended in the pseudopotentials, but you're right, making sure of that is important!

Regards,
Adam

**Reply by mesonepigreco** [reply]
Yes, fixed volume calculation just constrain the total volume instead of the target pressure, but relaxes the unit cell and the lattice anyway (unless it is cubic and fully constrained by symmetry). If you do not want to optimize the unit cell at all, you need to use the relax method instead of vc_relax.
Regarding the cutoff of the pseudo, be aware that the recommended values are usually tested to converge the total energy or the forces, note that usually the stress tensor requires a larger cutoff to get a precise result.

Bests,
Lorenzo

**Reply by adenchfi** [reply]
Hi Lorenzo,

I see! Yes, I am going above the pseudopotential recommended values, but I'm aware a proper convergence test must be done.

Your suggested fix for the pressure calculation worked! Thanks so much. Now my pressure doesn't increase without bound.

I will note I am still getting a segault with the same conditions as before - it prints AMASS {num1} {num2} then segfaults. Not sure what's going on still, and it seems like it's at the end of the procedure to save the H3S_minimizer results? Because it doesn't print out my "right before Hessian calc" statement, which is after the plot_results() command. And I would like to be able to save the Hessian in QE format for use in other codes. 

For testing purposes, is there a way to load the relax results from the run? To just go to the plot_results step and then the Hessian step?

**Reply by adenchfi** [reply]
PS: For posterity here's the new output from the H3S_minimizeroutput file. It has excellent convergence:
![image](https://user-images.githubusercontent.com/10607496/111849133-e0c22500-88da-11eb-8027-b5c48a0b11b6.png)

But yeah, would also like the rest of it. I get a core file from it, but the interface to analyze those seems challenging on a cluster.

**Reply by mesonepigreco** [reply]
Very glad to see it converges.

So it seems it crashes when computing the free energy hessian.
I open a new issue in the issues section to be fixed in the stable release. Can you tell us the FORTRAN compiler you use and the CPU? 
It seems that the hessian calculation has problems in some environments. We have a workaround. We provide a version of the code as a Docker container. This version has already been compiled correctly and should work independently on the machine used. In this way, you should be able to run the calculation even before we fix this problem.

The guide on how to use the docker container is here:
[www.sscha.eu/download](http://sscha.eu/download/)

section "Installation through Docker"

Now that you did the minimization, the code should have printed dynamical matrices in QE format as
dyn_popX_Y
where dyn_popX is the dynamical matrix after the population X. The last one is the converged one.
You can load the last ensemble with the last dynamical matrix to compute the Hessian, without the need for any other ab-initio calculation.

**Reply by adenchfi** [reply]
Hello Lorenzo,

Well, it seems to crash before I even call the Hessian module, based on:
![image](https://user-images.githubusercontent.com/10607496/111924823-4b5a9880-8a74-11eb-8ffe-4a92a731d612.png)

However, it could be that there's some cores or processors fetching ahead to the Hessian calculation and trying to start those calculations. 

I will try the docker version soon (probably tomorrow). Hopefully it runs fine on the cluster, will let you know.

> You can load the last ensemble with the last dynamical matrix to compute the Hessian, without the need for any other ab-initio calculation.

I see! I tried doing that, but it appears not all the data necessarily was saved. I tried your examples/ComputeFreeEnergyHessian approach of loading the ensemble, but it complains that "OSError: ./data_ensemble_autorelax/u_population100_1.dat not found.
'. 

There must be a flag to save that data that I didn't use, woops!  I am not sure what else to specify; I used 'save_ensemble=True' and 'Verbose=True' above. Is there another flag I'm unaware of? 

But rerunning the calculation shouldn't be a big deal, I'll just make sure it works with (much) smaller ensembles first :-) 

PS: I am using intel/17.04 icc compiler, and ifort as well. It is on the Bebop cluster at Argonne National Laboratory, on the bdw nodes, which have Intel Xeon E5-2695v4 processors.

**Reply by mesonepigreco** [reply]
Thanks for the specifications.
With intel compilers we had some problems, we will look more in details into this.

Anyway, for saving the ensembe you are doing it correctly. To load you have to use the load_bin function instead of load.

in fact the automatic relaxation saves the ensemble in binary files (which are faster to read and write and occupy less memory, but are not human readable).
These are saved and loaded with save_bin and load_bin
The load function, instead, suppose the ensemble is written in standard ASCII format, thus it looks for those files u_population100_X.dat files (the atomic displacements in ASCII format).

Try to load the ensemble with load_bin (specifying the correct population number). You may look to the directory  data_ensemble_autorelax to see which ensembles you have.

Bests,
Lorenzo

**Reply by adenchfi** [reply]
Hi Lorenzo,

Sorry for the delay. I had to wait for more compute time to be allocated, and a few other things got in the way. Thanks a ton for your assistance so far, it's been very helpful!

**Later edit: It seems to be working, except if one specifies only the gamma point k=[1,1,1] - see below.**

I've gotten the docker version working - our cluster uses udocker, but I managed to get it to work regardless. I successfully managed to use it to successfully compute/save the free energy Hessian for H3S. Next I wanted SSCHA to compute the spectral function, like in the PbTe example, using the SSCHA free energy hessian dynamical matrices (as they'd be the most accurate I presume). 

However, I get an error like so:

![image](https://user-images.githubusercontent.com/10607496/114341173-c0bf0380-9b1e-11eb-8ccd-bd4e6aff8c28.png)

I'm not sure if it's an nqirr symmetry thing or what. My function call is basically the same as in the PbTe example: 

![image](https://user-images.githubusercontent.com/10607496/114341297-0aa7e980-9b1f-11eb-9c49-259d2a1df62c.png)
with a reduced k-grid because even k = [4,4,4] produced hangups (I think it's just a memory issue, but I haven't figured out how to run SSCHA in parallel using docker on our slurm-managed cluster yet). I'm also not using the proper e1 param above since there are large phonon frequencies for H3S, but I don't think that would cause this(?).

The only other warning I see is earlier: 

![image](https://user-images.githubusercontent.com/10607496/114341586-c10bce80-9b1f-11eb-8645-ac71a9f79aa2.png)

and I'm not sure if this is relevant or not either. 

Finally:
In particular, is "nomm_spectral_func_[smear].dat" the final data file I want? that's the last thing that seems to successfully save. None of the other quantities after that (e.g. "v2_freq_shift_hwhm_perturb_[smear].dat
") seem to save to disk, and I'm not sure of their purpose. 

PS: I reran the above with k=[1,1,2] and the proper e1 upper bound. It seemed to run fine. Maybe it's just a k=[1,1,1] edge case bug.

**Reply by mesonepigreco** [reply]
The most important files are nomm_spectral_func_[smear].dat, which stores the spectral functions. Then the code extracts from the spectral function useful info like the frequencies and lifetimes in the Lorentzian approximation. The most relevant info is, however, inside the spectral function. This bug seems to take place in this last part of the code, so it does not affect the spectral function itself. So:

> In particular, is "nomm_spectral_func_[smear].dat" the final data file I want?

Yes!

@rafbianco @marcocherubini 
Have you seen this issue before? I remember something like that, but I thought we fixed it.
Can you have a look at it?

**Reply by marcocherubini** [reply]
Hi, 
It seems like the phonon frequencies you get are far outside the energy range you are setting.
If I understood correctly, you said that you are using the SSCHA free energy hessian dynamical matrices as the input for get_diag_dynamic_correction_along_path, but you should use the SSCHA dynamical matrices (the ones you get after the minimization).
In fact, in the hessian matrices you are already including the static correction. Basically, you are applying twice the correction
(even if the second time in the dynamical approximation).
This is for sure a source of error, even if I don't think it is the reason of the error you are getting. The static correction reduces the frequencies, so, it should not push them away from the energy range. Moreover, I think it is strange that in a bigger grid the bug disappears, since in any case the [1,1,2] contains the [1,1,1] where you have a problem. If you think the phonon frequencies you get are reasonable, you should increase the energy range (e0,e1) and this should solve the problem.

P.S. For what concerns the memory issue, we had something similar, but now it is fixed (in the master).
So, it is strange you cannot use a bigger k-grid, the memory cost is the same as for the [1,1,1].

**Reply by adenchfi** [reply]
Thanks you both! 

> Yes!
Excellent. I plotted the spectral function and it seems fine, very interesting to look at. 

> It seems like the phonon frequencies you get are far outside the energy range you are setting.

Yep, I've fixed that now.

> In fact, in the hessian matrices you are already including the static correction. Basically, you are applying twice the correction

Okay, good to know I should use the SSCHA minimized quantities and not the free energy hessian. I'll need to review the SSCHA papers again now that I've actually used the software :-) 

> P.S. For what concerns the memory issue, we had something similar, but now it is fixed (in the master).

Got it. To get it working on our cluster I'm using the docker version (as per above discussion); does the current docker link reflect these changes?

Unrelated but, what format is the q-path file expecting? It doesn't look like crystal coordinates. I thought it was 2pi/alat, but I'm not sure if the XGX_path file in the PbTe example is reflecting that (or my own ignorance is again shining).

**Reply by rafbianco** [reply]
Hi,
the expected q-path format is 2pi/Angstrom.

**Reply by mesonepigreco** [reply]
> Got it. To get it working on our cluster I'm using the docker version (as per above discussion); does the current docker link reflect these changes?


Not yet, I will prepare a new docker soon once we release the next alpha version of the code together with an update of the pip python packages. I am waiting to collect more bugfixes together.


@rafbianco 

>Okay, good to know I should use the SSCHA minimized quantities and not the free energy hessian. 

Maybe we should change the tutorials on the spectral function specifying this more clearly with an alert, bold text or something like that, it seems a dangerous issue; something people could misunderstand.
